flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java

```diff
@@ -1329,10 +1331,13 @@ private void discardCheckpoint(PendingCheckpoint pendingCheckpoint, @Nullable Th
 
 		LOG.info("Discarding checkpoint {} of job {}.", checkpointId, job, cause);
 
-		if (cause == null || cause instanceof CheckpointDeclineException) {
-			pendingCheckpoint.abort(CheckpointFailureReason.CHECKPOINT_DECLINED, cause);
+		if (cause == null) {
+			failPendingCheckpoint(pendingCheckpoint, CheckpointFailureReason.CHECKPOINT_DECLINED);
+		} else if (cause instanceof CheckpointException) {
+			CheckpointException exception = (CheckpointException) cause;
+			failPendingCheckpoint(pendingCheckpoint, exception.getCheckpointFailureReason(), cause);
 		} else {
-			pendingCheckpoint.abort(CheckpointFailureReason.JOB_FAILURE, cause);
+			failPendingCheckpoint(pendingCheckpoint, CheckpointFailureReason.JOB_FAILURE, cause);
```

Comment:
I noticed that you actually did not touch the `AbstractStreamOperator` part. With your PR, if operator fail to complete the snapshot in method `#snapshotState(long , long , CheckpointOptions , CheckpointStreamFactory )` and it would only decline that checkpoint and return as an `Exception`. In other words, `CheckpointFailureManager` would only process the failed checkpoint in this line which results in not increasing `continuousFailureCounter` in `CheckpointFailureManager`. Is this reasonable?

Reply:
Actually, `CheckpointFailureManager` can take more effect in the future but not now. This is an intermediate step of the whole three PRs. In this PR, we need to keep compatible with `setFailOnCheckpointingErrors`, it's the most important thing. Otherwise, it will change many user behaviors. We have considered counting more failure reason before, but it will make more changes and make this PR more complex. So your thought is right but not for now.

The purpose of this PR is to introduce the `CheckpointFailureManager` and do further residual refactor work for the first PR #7571.